src/fastify-plugins/maps/index.js

```diff
@@ -20,57 +24,153 @@ export const plugin = fp(mapsPlugin, {
-
-    // 1. Attempt to get "default" local static map's style.json
-    {
-      const styleId = 'default'
-
-      const results = await Promise.all([
-        fastify.mapeoStaticMaps.getStyleJsonStats(styleId),
-        fastify.mapeoStaticMaps.getResolvedStyleJson(styleId, serverAddress),
-      ]).catch(() => {
-        fastify.log.warn('Cannot read default static map')
-        return null
-      })
-
-      if (results) {
+async function routes(fastify, opts) {
+  const { defaultOnlineStyleUrl = DEFAULT_MAPBOX_STYLE_URL } = opts
+
+  fastify.get(
+    '/style.json',
+    { schema: { querystring: GetStyleJsonQueryStringSchema } },
+    async (req, rep) => {
+      const serverAddress = await getFastifyServerAddress(req.server.server)
+
+      // 1. Attempt to get "default" local static map's style.json
+      {
+        const styleId = 'default'
+
+        const results = await Promise.all([
+          fastify.mapeoStaticMaps.getStyleJsonStats(styleId),
+          fastify.mapeoStaticMaps.getResolvedStyleJson(styleId, serverAddress),
+        ]).catch(() => {
+          fastify.log.warn('Cannot read default static map')
+          return null
+        })
+
+        if (results) {
+          const [stats, styleJson] = results
+          rep.headers(createStyleJsonResponseHeaders(stats.mtime))
+          return styleJson
+        }
+      }
+
+      // 2. Attempt to get a default style.json from online source
+      {
+        const { key } = req.query
+        const upstreamUrl = key
+          ? styleUrlWithApiKey(defaultOnlineStyleUrl, key)
+          : defaultOnlineStyleUrl
+
+        try {
+          const upstreamResponse = await fetch(upstreamUrl)
+
+          if (upstreamResponse.ok) {
+            // Set up headers to forward
+            for (const [name, value] of upstreamResponse.headers) {
+              // TODO: Typically Mapbox gzips the content but we'd need to make sure
+              // something like https://github.com/fastify/fastify-compress/ is registered
+              if (name.toLowerCase() === 'content-encoding') continue
+              rep.header(name, value)
+            }
+            // Some upstream providers will not set the 'application/json' content-type header despite the body being JSON e.g. Protomaps
+            // TODO: Should we forward the upstream 'content-type' header?
+            // We kind of assume that a Style Spec-compatible JSON payload will always be used by a provider
+            // Tehcnically, there could be cases where a provider doesn't use the Mapbox Style Spec and has their own format,
+            // which may be delivered as some other content type
+            rep.header('content-type', 'application/json; charset=utf-8')
+            return upstreamResponse.json()
+          } else {
+            fastify.log.error(
+              `Upstream style.json request failed: ${upstreamResponse.statusText}`
+            )
+          }
+        } catch (err) {
+          fastify.log.error(err)
```

Comment:
1. I think this should log a warning, not an error. No developer has made a mistake if this request fails.
2. Should we add a little info about what went wrong?

```suggestion
          fastify.log.error('Upstream style.json request totally failed:', err)
```

Reply:
initial thought is that it's a little too similar in wording to the nested message about getting a bad response. maybe one of these should be reworded to have more differentiation?

Reply:
Yeah...I think there's a "network error" case and a "non-2XX error". Maybe we update the log messages with that kind of thing?

Reply:
addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/0654550d7c90f2cbaf3368421b97853324decf15